compiler/rustc_trait_selection/src/traits/normalize.rs

```diff
@@ -178,6 +176,176 @@ impl<'a, 'b, 'tcx> AssocTypeNormalizer<'a, 'b, 'tcx> {
+            // after normalization succeeded.
+            //
+            // Also, as an optimization: when we don't have escaping bound vars, we don't
+            // need to replace them with placeholders (see branch below).
+            let proj = proj.fold_with(self);
+            let normalized_term = project::normalize_projection_term(
+                self.selcx,
+                self.param_env,
+                proj,
+                self.cause.clone(),
+                self.depth,
+                self.obligations,
+            );
+            debug!(
+                ?self.depth,
+                ?proj,
+                ?normalized_term,
+                obligations.len = ?self.obligations.len(),
+                "AssocTypeNormalizer: normalized term"
+            );
+            normalized_term
+        } else {
+            // If there are escaping bound vars, we temporarily replace the
+            // bound vars with placeholders. Note though, that in the case
+            // that we still can't project for whatever reason (e.g. self
+            // type isn't known enough), we *can't* register an obligation
+            // and return an inference variable (since then that obligation
+            // would have bound vars and that's a can of worms). Instead,
+            // we just give up and fall back to pretending like we never tried!
+            //
+            // Note: this isn't necessarily the final approach here; we may
+            // want to figure out how to register obligations with escaping vars
+            // or handle this some other way.
+
+            let infcx = self.selcx.infcx;
+            let orig_proj = proj;
+            let (proj, mapped_regions, mapped_types, mapped_consts) =
+                BoundVarReplacer::replace_bound_vars(infcx, &mut self.universes, proj);
+            let proj = proj.fold_with(self);
+            let normalized_term = project::opt_normalize_projection_term(
+                self.selcx,
+                self.param_env,
+                proj,
+                self.cause.clone(),
+                self.depth,
+                self.obligations,
+            )
+            .ok()
+            .flatten()
+            .map(|normalized_term| {
+                PlaceholderReplacer::replace_placeholders(
+                    infcx,
+                    mapped_regions,
+                    mapped_types,
+                    mapped_consts,
+                    &self.universes,
+                    normalized_term,
+                )
+            })
+            // We fold the original projection term before its bound vars were replaced with placeholders
+            // as otherwise we'd leak placeholders from normalization. This `fold_with` is also not a trivial
+            // recursion/cycle as folding over an `AliasTerm` just folds over the generic arguments to the alias.
+            .unwrap_or_else(|| orig_proj.fold_with(self).to_term(infcx.tcx));
```

Comment:
why not move this above the `replace_placeholders` instead?

Reply:
:thinking: yeah thats probably less confusing to do the folding and to_term earlier

Reply:
it makes the comment kinda awkward actually because we've not yet failed to normalize :thinking:

Reply:
nvm :3 I think I understand what you meant? This is a lot nicer if it's what you meant